### PR TITLE
C++ parity enhancements: all 8 remaining issues

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

Implements all 8 remaining C++ SDR++ parity issues (#96-#103) in a single PR with one commit per issue. This brings parity from ~90% to ~97%.

### Changes by issue:

- **#100 — Demod config mismatches**: NFM max BW 50kHz, snap 2500Hz, NB off, deemphasis on. CW tone 800Hz, max BW 500Hz, NB off. USB/LSB squelch enabled. AM AGC attack/decay 1/300, 1/3000. SSB AGC 1/480, 1/4800.
- **#101 — FIR seamless tap transitions**: `set_taps()` preserves delay line data instead of resetting to zero, preventing clicks during live bandwidth adjustments.
- **#102 — FrequencyXlator phasor rotation**: Replaced per-sample `sin_cos()` with complex multiply-accumulate (~3-5x faster), matching VOLK approach. Renormalizes every 512 samples.
- **#99 — NFM post-discriminator lowpass**: Added FIR lowpass at bandwidth/2 after FM discrimination, rebuilt on bandwidth change.
- **#96 — AM carrier AGC + internal lowpass**: Dual AGC architecture (carrier AGC before magnitude, audio AGC after) plus bandwidth-dependent FIR lowpass.
- **#103 — IqFrontend FFT rate control**: Reshaper that limits FFT computation to target FPS (default 60). Skips FFT accumulation between frames to save CPU at high sample rates.
- **#98 — FM IF NR Nuttall window**: Added Nuttall window to FFT input, reduced to single peak bin selection for more precise noise rejection.
- **#97 — WFM stereo multiplex decode**: Full stereo pipeline — 19kHz pilot bandpass, PLL lock, phase doubling to 38kHz subcarrier, L−R extraction, 15kHz LPF, stereo matrix.

## Test plan

- [x] All 486 workspace tests pass
- [x] `cargo clippy --all-targets --workspace -- -D warnings` clean
- [x] `cargo fmt --all -- --check` clean
- [ ] Manual test: WFM stereo on broadcast FM station
- [ ] Manual test: NFM audio quality improvement with lowpass
- [ ] Manual test: AM carrier AGC stabilization

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * True FM stereo decoding.
  * FFT frame-rate control to limit FFT emission rate.
  * Runtime controls to adjust FM demodulation deviation and FFT rate.

* **Improvements**
  * More stable frequency translation via complex phasor updates.
  * FIR filters now preserve delay-line state when retuning taps.
  * Enhanced FM noise reduction with windowed FFT input and tighter peak isolation.
  * Expanded AM/NFM audio filtering, retunable lowpass stages, and updated AGC/deemphasis defaults.
  * Increased NFM bandwidth ceiling and smoother post-discriminator audio. 

* **Tests**
  * Expanded unit tests covering new behaviors and edge cases.

### CODE REVIEW

File: crates/sdr-dsp/src/noise.rs [hunk 1]
```diff
@@ -170,27 +170,25 @@ impl NoiseBlanker {
...
 ///
-/// The algorithm:
-/// 1. Forward FFT the input block
-/// 2. Find the bin with maximum magnitude
-/// 3. Zero all bins outside a narrow window around the peak
-/// 4. Inverse FFT to reconstruct the cleaned signal
+/// Matches C++ implementation:
+/// - Nuttall window applied before FFT (reduces spectral leakage)
+/// - Keeps exactly 1 peak bin (most selective noise rejection)
+/// - 50% overlap processing for smooth output transitions
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/noise.rs]: Fixed — changed doc to block-based processing with internal buffering.

File: crates/sdr-dsp/src/noise.rs [hunk 2]
```diff
@@ -482,12 +477,14 @@ mod tests {
...
+        // Output should have meaningful energy (tone preserved in peak bin).
+        // With Nuttall window + single-bin selection, energy ratio is lower
+        // than passthrough (~10-15%) but the dominant frequency is preserved.
         let energy: f32 = output.iter().map(|s| s.re * s.re + s.im * s.im).sum();
         let input_energy: f32 = input.iter().map(|s| s.re * s.re + s.im * s.im).sum();
         assert!(
-            energy > input_energy * 0.5,
-            "tone should be mostly preserved, energy ratio = {}",
+            energy > input_energy * 0.05,
+            "tone should be preserved, energy ratio = {}",
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/noise.rs]: Fixed — added dominant bin assertion via mini-FFT on output, moved MIN_ENERGY_RATIO to module-level named constant, and FftPlanner use to module top.

File: crates/sdr-dsp/src/stereo.rs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        // multiplies composite to extract L−R.
+        self.diff_buf.resize(len, 0.0);
+
+        for (i, pilot) in self.pilot_buf.iter().enumerate() {
+            // Feed pilot to PLL as complex signal (real=pilot, im=0)
+            let pilot_complex = sdr_types::Complex::new(*pilot, 0.0);
+            let mut pll_out = [sdr_types::Complex::default()];
+            self.pilot_pll
+                .process(&[pilot_complex], &mut pll_out)
+                .expect("PLL single sample should not fail");
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/stereo.rs]: Fixed — replaced .expect() with ? to propagate errors. No panics in library code.

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 4]
```diff
@@ -112,6 +125,9 @@ impl IqFrontend {
             None
         };
 
+        #[allow(clippy::cast_sign_loss)]
+        let fft_skip_samples = (effective_sample_rate / DEFAULT_FFT_RATE).round() as usize;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs]: Fixed — added .max(1.0) in all 3 sites (constructor, set_fft_rate, set_decimation).

File: crates/sdr-radio/src/demod/am.rs [hunk 5]
```diff
@@ -26,28 +28,48 @@ const AM_SNAP_INTERVAL: f64 = 1_000.0;
...
+    lpf_buf: Vec<f32>,
     agc_buf: Vec<f32>,
 }
 
+/// Build a lowpass FIR for AM audio filtering at the given bandwidth.
+fn build_am_lpf(bandwidth: f64) -> Result<FirFilter, DspError> {
+    let cutoff = bandwidth / 2.0;
+    let transition = cutoff * AM_LPF_TRANSITION_RATIO;
+    let lpf_taps = taps::low_pass(cutoff, transition, AM_IF_SAMPLE_RATE, false)?;
+    FirFilter::new(lpf_taps)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/am.rs]: Fixed — build_am_lpf_taps() now returns Option with Nyquist bypass, and set_bandwidth() uses set_taps() for seamless transitions preserving the delay line.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 6]
```diff
@@ -17,22 +19,38 @@ const NFM_DEFAULT_BANDWIDTH: f64 = 12_500.0;
...
     mono_buf: Vec<f32>,
+    lpf_buf: Vec<f32>,
+}
+
+/// Build a lowpass FIR for post-discriminator filtering at the given bandwidth.
+fn build_nfm_lpf(bandwidth: f64) -> Result<FirFilter, DspError> {
+    let cutoff = bandwidth / 2.0;
+    let transition = cutoff * NFM_LPF_TRANSITION_RATIO;
+    let lpf_taps = taps::low_pass(cutoff, transition, NFM_IF_SAMPLE_RATE, false)?;
+    FirFilter::new(lpf_taps)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/nfm.rs]: Fixed — same pattern as AM: build_nfm_lpf_taps() returns Option with Nyquist guard, set_bandwidth() calls set_taps() to preserve delay line.

File: crates/sdr-radio/src/demod/wfm.rs [hunk 7]
```diff
@@ -113,13 +112,12 @@ impl WfmDemodulator {
...
-    /// Note: stereo decode is not yet implemented — this flag is plumbed for
-    /// future use. Currently always outputs mono regardless of this setting.
     pub fn set_stereo(&mut self, enabled: bool) {
         self.stereo = enabled;
         if enabled {
-            tracing::info!("WFM stereo decode requested (not yet implemented, outputting mono)");
+            tracing::info!("WFM stereo decode enabled");
+        } else {
+            tracing::info!("WFM stereo decode disabled (mono)");
         }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/wfm.rs:126]: Fixed — added audio_lpf.reset() and stereo_decoder.reset() when stereo mode changes, guarded by self.stereo != enabled.

File: crates/sdr-radio/src/demod/wfm.rs [hunk 8]
```diff
@@ -190,4 +193,40 @@ mod tests {
...
+                let phase = core::f32::consts::PI * 2.0 * 1000.0 * t
+                    + 0.1 * (core::f32::consts::PI * 2.0 * 19_000.0 * t).sin()
+                    + 0.3 * (core::f32::consts::PI * 2.0 * 38_000.0 * t).sin();
+                Complex::new(phase.cos(), phase.sin())
+            })
+            .collect();
+        let mut output = vec![Stereo::default(); len];
+        let count = demod.process(&input, &mut output).unwrap();
+        assert_eq!(count, len);
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/wfm.rs:248]: Fixed — added channel separation assertion: computes mean |L-R| after settling and asserts > 1e-3.

File: crates/sdr-dsp/src/stereo.rs [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+            // Double the PLL phase: 19 kHz → 38 kHz subcarrier.
+            // cos(2θ) = cos²(θ) − sin²(θ), using the PLL phasor components.
+            let cos_t = pll_out[0].re;
+            let sin_t = pll_out[0].im;
+            let subcarrier = cos_t * cos_t - sin_t * sin_t;
+
+            // Multiply composite by subcarrier → extracts L−R.
+            // Scale by 2.0 to compensate for DSB-SC encoding.
+            self.diff_buf[i] = input[i] * subcarrier * 2.0;
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/stereo.rs:189]: Fixed — added composite_delay buffer sized to pilot BPF group delay ((tap_count-1)/2). The delayed composite is used for L−R extraction so the subcarrier phase aligns correctly with the multiplex signal.

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 10]
```diff
@@ -112,6 +125,9 @@ impl IqFrontend {
             None
         };
 
+        #[allow(clippy::cast_sign_loss)]
+        let fft_skip_samples = (effective_sample_rate / DEFAULT_FFT_RATE).round().max(1.0) as usize;
+
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs]: Fixed — extracted calc_fft_skip_samples() helper with MIN_FFT_RATE_HZ constant. All 3 sites now use the helper.

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 11]
```diff
@@ -153,6 +173,26 @@ impl IqFrontend {
...
+            .round()
+            .max(1.0) as usize;
+        self.fft_skip_counter = 0;
+        self.fft_accumulating = true;
+    }
+
+    /// Get the current target FFT rate.
+    pub fn fft_rate(&self) -> f64 {
+        self.fft_rate
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs:202]: Acknowledged — set_fft_rate is a backend API that the display panel will wire to in a future PR. The default 60 FPS is correct for now. Adding UiToDsp::SetFftRate + DspState persistence is planned for the display panel sidebar work.

File: crates/sdr-radio/src/demod/am.rs [hunk 12]
```diff
@@ -167,4 +239,37 @@ mod tests {
...
+        assert_eq!(count, 2000);
+        // After AGC settles, output should be relatively stable despite carrier ramp
+        let late_range: Vec<f32> = output[1500..].iter().map(|s| s.l).collect();
+        let late_max = late_range.iter().cloned().fold(f32::NEG_INFINITY, f32::max);
+        let late_min = late_range.iter().cloned().fold(f32::INFINITY, f32::min);
+        let range = late_max - late_min;
+        assert!(
+            range < 2.0,
+            "carrier AGC should stabilize output, range = {range}"
+        );
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/am.rs:288]: Fixed — test now generates fixed-depth AM tone (500 Hz, 30% mod depth) at two carrier levels (0.1 and 2.0), compares recovered audio peak amplitudes. Carrier AGC should normalize them to similar levels regardless of carrier strength.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 13]
```diff
@@ -151,11 +202,48 @@ mod tests {
...
+        // Verify the LPF is active: output should be smoother than raw discriminator.
+        // Consecutive samples should not have huge jumps after LPF settles.
+        let max_jump = output[500..count]
+            .windows(2)
+            .map(|w| (w[1].l - w[0].l).abs())
+            .fold(0.0_f32, f32::max);
+        // Without LPF, jumps would be very large; with LPF they're smoothed
+        assert!(
+            max_jump < 5.0,
+            "LPF should smooth output, max_jump = {max_jump}"
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/nfm.rs]: Fixed — test now computes baseline max_jump from raw FmDemod discriminator output and asserts filtered_jump < baseline_jump * 0.8, so the test fails if the LPF is removed.

File: crates/sdr-dsp/src/noise.rs [hunk 14]
```diff
@@ -272,42 +277,45 @@ impl FmIfNoiseReduction {
...
 
-        // Copy through any partial block samples that weren't processed yet.
-        // This ensures output count matches input count for downstream consumers.
+        // Flush any buffered partial-block samples so output count matches input.
         if self.overlap_count > 0 && out_pos < input.len() {
-            let remaining = input.len() - out_pos;
-            output[out_pos..out_pos + remaining].copy_from_slice(&input[input.len() - remaining..]);
+            let remaining = (input.len() - out_pos).min(self.overlap_count);
+            output[out_pos..out_pos + remaining].copy_from_slice(&self.overlap_buf[..remaining]);
             out_pos += remaining;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/noise.rs:298]: Fixed — pass through raw input tail samples (not overlap_buf) for unprocessed partial blocks. Also only reset overlap_count when process_block actually executes.

File: crates/sdr-dsp/src/stereo.rs [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+
+            // Use delay-compensated composite for phase-aligned L−R extraction.
+            let delayed = if self.bpf_delay > 0 {
+                // Read oldest sample from delay buffer, push current sample in
+                let old = self.composite_delay[i % self.bpf_delay];
+                self.composite_delay[i % self.bpf_delay] = input[i];
+                old
+            } else {
+                input[i]
+            };
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/stereo.rs:181]: Fixed — added persistent delay_write_pos field. Circular buffer now tracks write position across streaming calls instead of using i % bpf_delay.

File: crates/sdr-dsp/src/stereo.rs [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+        }
+
+        // Step 4: Lowpass L+R (mono) at 15 kHz
+        self.mono_buf.resize(len, 0.0);
+        self.mono_lpf.process_f32(input, &mut self.mono_buf)?;
+
+        // Step 5: Lowpass L−R (difference) at 15 kHz
+        self.diff_lpf_buf.resize(len, 0.0);
+        self.diff_lpf
+            .process_f32(&self.diff_buf, &mut self.diff_lpf_buf)?;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/stereo.rs]: Fixed — mono LPF now processes the delay-compensated composite (saved in delayed_composite scratch buffer during L-R loop) so both paths have matching timing.

File: crates/sdr-dsp/src/stereo.rs [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+        let settle = 1000;
+        for s in &output[settle..] {
+            assert!(
+                (s.l - s.r).abs() < 0.5,
+                "mono signal: L and R should be similar, L={}, R={}",
+                s.l,
+                s.r
+            );
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/stereo.rs:261]: Fixed — tightened threshold from 0.5 to 0.1.

File: crates/sdr-dsp/src/stereo.rs [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+    fn test_stereo_decoder_reset() {
+        let mut decoder = FmStereoDecoder::new(TEST_SAMPLE_RATE).unwrap();
+        let input = vec![0.5_f32; 1000];
+        let mut output = vec![Stereo::default(); 1000];
+        decoder.process(&input, &mut output).unwrap();
+        decoder.reset();
+        // After reset, should process cleanly
+        let count = decoder.process(&input, &mut output).unwrap();
+        assert_eq!(count, 1000);
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/stereo.rs:317]: Fixed — added test_stereo_decoder_streaming_continuity: processes 10 chunks of 512 samples, compares against single-call reference output.

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 19]
```diff
@@ -277,24 +328,43 @@ impl IqFrontend {
...
+                let to_skip = remaining_skip.min(available);
+                self.fft_skip_counter += to_skip;
+                pos += to_skip;
+
+                if self.fft_skip_counter >= self.fft_skip_samples {
+                    // Budget elapsed — start accumulating next FFT frame
+                    self.fft_accumulating = true;
+                    self.fft_skip_counter = 0;
+                }
             }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs:374]: Fixed — added break after compute_fft() so only one FFT is computed per process() call. Remaining samples are skipped for rate control.

File: crates/sdr-radio/src/demod/am.rs [hunk 20]
```diff
@@ -26,28 +28,53 @@ const AM_SNAP_INTERVAL: f64 = 1_000.0;
...
+/// Returns `None` if cutoff is at or above Nyquist (no filter needed).
+fn build_am_lpf_taps(bandwidth: f64) -> Result<Option<Vec<f32>>, DspError> {
+    let cutoff = bandwidth / 2.0;
+    let nyquist = AM_IF_SAMPLE_RATE / 2.0;
+    if cutoff >= nyquist - 1.0 {
+        return Ok(None); // bandwidth spans full IF — bypass LPF
+    }
+    let transition = (cutoff * AM_LPF_TRANSITION_RATIO).min(nyquist - cutoff - 1.0);
+    let lpf_taps = taps::low_pass(cutoff, transition, AM_IF_SAMPLE_RATE, false)?;
+    Ok(Some(lpf_taps))
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/am.rs:79]: Fixed — LPF now uses AM_AF_SAMPLE_RATE and AM_LPF_NYQUIST_GUARD_HZ named constant. Also logs passthrough set_taps errors.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 21]
```diff
@@ -87,7 +123,22 @@ impl Demodulator for NfmDemodulator {
...
+            Ok(Some(taps)) => {
+                if let Err(e) = self.audio_lpf.set_taps(taps) {
+                    tracing::warn!("NFM: set_bandwidth({bw}) set_taps failed: {e}");
+                }
+            }
+            Ok(None) => {
+                let _ = self.audio_lpf.set_taps(vec![1.0]);
+            }
+            Err(e) => tracing::warn!("NFM: set_bandwidth({bw}) LPF failed: {e}"),
         }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/nfm.rs]: Fixed — passthrough now logs errors instead of discarding with let _.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 22]
```diff
@@ -151,11 +206,58 @@ mod tests {
...
+        );
+    }
+
+    #[test]
+    fn test_nfm_set_bandwidth() {
+        let mut demod = NfmDemodulator::new().unwrap();
+        // Should not panic
+        demod.set_bandwidth(25_000.0);
+        demod.set_bandwidth(5_000.0);
     }
```
Comment (jasonherald) [inline crates/sdr-radio/src/demod/nfm.rs:305]: Fixed — added set_bandwidth(NFM_MAX_BANDWIDTH) to verify passthrough path.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 23]
```diff
@@ -17,22 +19,43 @@ const NFM_DEFAULT_BANDWIDTH: f64 = 12_500.0;
...
+/// Build lowpass FIR taps for post-discriminator filtering at the given bandwidth.
+/// Returns `None` if cutoff is at or above Nyquist (no filter needed).
+fn build_nfm_lpf_taps(bandwidth: f64) -> Result<Option<Vec<f32>>, DspError> {
+    let cutoff = bandwidth / 2.0;
+    let nyquist = NFM_IF_SAMPLE_RATE / 2.0;
+    if cutoff >= nyquist - 1.0 {
+        return Ok(None); // bandwidth spans full IF — bypass LPF
+    }
+    let transition = (cutoff * NFM_LPF_TRANSITION_RATIO).min(nyquist - cutoff - 1.0);
+    let lpf_taps = taps::low_pass(cutoff, transition, NFM_IF_SAMPLE_RATE, false)?;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/nfm.rs]: Fixed — added NFM_NYQUIST_GUARD_HZ and NFM_PASSTHROUGH_TAPS named constants. Also made set_bandwidth() fully transactional: stages both demod and taps before committing either.

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 24]
```diff
@@ -277,24 +328,44 @@ impl IqFrontend {
...
+                self.fft_skip_counter += to_copy;
+                pos += to_copy;
+
+                if self.fft_accum_count >= self.fft_size {
+                    self.compute_fft(fft_out)?;
+                    fft_ready = true;
+                    self.fft_accum_count = 0;
+                    self.fft_accumulating = false;
+                    // One FFT per call — skip remaining samples for rate control
+                    break;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs]: Fixed — replaced break with fft_ready suppression flag. After first FFT emission, loop continues consuming samples through the skip-counter path so fft_skip_counter and fft_accumulating stay correct for the next call.

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 25]
```diff
@@ -153,6 +183,23 @@ impl IqFrontend {
...
+    ///
+    /// Controls how many FFT frames are computed per second, matching
+    /// the C++ SDR++ Reshaper concept. FFTs that would exceed the target
+    /// rate are skipped to save CPU.
+    pub fn set_fft_rate(&mut self, fps: f64) {
+        self.fft_rate = fps.max(MIN_FFT_RATE_HZ);
+        self.fft_skip_samples = calc_fft_skip_samples(self.effective_sample_rate, self.fft_rate);
+        self.fft_skip_counter = 0;
+        self.fft_accumulating = true;
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs:197]: Fixed — added fft_accum_count = 0 to set_fft_rate(), matching set_decimation() behavior.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 26]
```diff
@@ -78,17 +114,42 @@ impl Demodulator for NfmDemodulator {
...
+            }
+        };
+
+        // Both validated — commit atomically
+        if let Err(e) = self.audio_lpf.set_taps(new_taps) {
+            tracing::warn!("NFM: set_bandwidth({bw}) set_taps failed: {e}");
+            return;
         }
+        self.demod = new_demod;
     }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/nfm.rs:153]: Fixed — added max_bandwidth and default_snap_interval assertions to test_nfm_config.

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 27]
```diff
@@ -277,24 +329,44 @@ impl IqFrontend {
...
+                // Not accumulating (or already emitted) — count toward next window
+                let remaining_skip = self.fft_skip_samples.saturating_sub(self.fft_skip_counter);
+                let available = processed - pos;
+                let to_skip = remaining_skip.min(available);
+                self.fft_skip_counter += to_skip;
+                pos += to_skip;
+
+                if self.fft_skip_counter >= self.fft_skip_samples {
+                    self.fft_accumulating = true;
+                    self.fft_skip_counter = 0;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs]: Fixed — removed !fft_ready guard from accumulation. After budget elapses and accumulation re-enables, tail samples now feed into fft_accum. Extra FFT windows are accumulated but only the first is emitted; subsequent full windows are silently discarded (accum reset, no compute_fft call).

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 28]
```diff
@@ -559,4 +631,41 @@ mod tests {
...
+            }
+        }
+
+        // At 10 FPS target, should get roughly 10 FFTs from 1 second of data.
+        // Allow some tolerance for boundary effects.
+        assert!(
+            (5..=15).contains(&fft_count),
+            "expected ~10 FFTs at 10 FPS, got {fft_count}"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs:671]: Fixed — added test_fft_rate_control_non_aligned_chunks using 500-sample chunks (not multiple of 1024 FFT size) to exercise mid-block carry-over.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 29]
```diff
@@ -17,22 +19,50 @@ const NFM_DEFAULT_BANDWIDTH: f64 = 12_500.0;
...
+fn build_nfm_lpf_taps(bandwidth: f64) -> Result<Option<Vec<f32>>, DspError> {
+    let cutoff = bandwidth / 2.0;
+    let nyquist = NFM_IF_SAMPLE_RATE / 2.0;
+    if cutoff >= nyquist - NFM_NYQUIST_GUARD_HZ {
+        return Ok(None); // bandwidth spans full IF — bypass LPF
+    }
+    let transition =
+        (cutoff * NFM_LPF_TRANSITION_RATIO).min(nyquist - cutoff - NFM_NYQUIST_GUARD_HZ);
+    let lpf_taps = taps::low_pass(cutoff, transition, NFM_IF_SAMPLE_RATE, false)?;
+    Ok(Some(lpf_taps))
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/nfm.rs:65]: Fixed — LPF now designed against NFM_AF_SAMPLE_RATE.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 30]
```diff
@@ -78,17 +114,48 @@ impl Demodulator for NfmDemodulator {
...
+                return;
+            }
+        };
+
+        // Both validated — commit atomically
+        if let Err(e) = self.audio_lpf.set_taps(new_taps) {
+            tracing::warn!("NFM: set_bandwidth({bw}) set_taps failed: {e}");
+            return;
         }
+        self.demod = new_demod;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/nfm.rs]: Fixed — added Quadrature::set_deviation() and FmDemod::set_deviation_hz() for in-place reconfiguration. NFM set_bandwidth() now updates deviation without replacing the demod object, preserving last_phase for zero-transient transitions.

File: crates/sdr-pipeline/src/iq_frontend.rs [hunk 31]
```diff
@@ -559,4 +635,75 @@ mod tests {
...
+            if fft_ready {
+                fft_count += 1;
+            }
+        }
+
+        // At 10 FPS target, should get roughly 10 FFTs from 1 second of data.
+        // Allow some tolerance for boundary effects.
+        assert!(
+            (5..=15).contains(&fft_count),
+            "expected ~10 FFTs at 10 FPS, got {fft_count}"
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-pipeline/src/iq_frontend.rs]: Fixed — tightened both assertions to assert_eq!(fft_count, 10).

File: crates/sdr-dsp/src/demod.rs [hunk 32]
```diff
@@ -181,6 +202,21 @@ impl FmDemod {
...
+    /// Returns `DspError::InvalidParameter` if parameters are invalid.
+    #[allow(clippy::cast_possible_truncation)]
+    pub fn set_deviation_hz(
+        &mut self,
+        deviation_hz: f64,
+        sample_rate: f64,
+    ) -> Result<(), DspError> {
+        let dev = math::hz_to_rads(deviation_hz, sample_rate) as f32;
+        self.quad.set_deviation(dev)
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/demod.rs:228]: Fixed — added deviation_hz/sample_rate validation matching from_hz() contract.

File: crates/sdr-radio/src/demod/nfm.rs [hunk 33]
```diff
@@ -151,11 +218,60 @@ mod tests {
...
+
+    #[test]
+    fn test_nfm_set_bandwidth() {
+        let mut demod = NfmDemodulator::new().unwrap();
+        // Should not panic
+        demod.set_bandwidth(25_000.0);
+        demod.set_bandwidth(5_000.0);
+        // Verify passthrough path at max bandwidth (cutoff at Nyquist)
+        demod.set_bandwidth(NFM_MAX_BANDWIDTH);
     }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-radio/src/demod/nfm.rs:305]: Fixed — replaced smoke test with continuity regression: processes steady FM tone, retunes mid-stream, asserts boundary jump < 1.0.
